src/picongpu/include/plugins/ChargeConservation.tpp

```diff
@@ -0,0 +1,156 @@
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "math/vector/Int.hpp"
+#include "math/vector/Float.hpp"
+#include "math/vector/Size_t.hpp"
+#include "math/vector/math_functor/abs.hpp"
+#include "math/vector/math_functor/max.hpp"
+#include "cuSTL/container/PseudoBuffer.hpp"
+#include "dataManagement/DataConnector.hpp"
+#include "fields/FieldJ.hpp"
+#include "math/Vector.hpp"
+#include "cuSTL/algorithm/mpi/Gather.hpp"
+#include "cuSTL/container/DeviceBuffer.hpp"
+#include "cuSTL/container/HostBuffer.hpp"
+#include "cuSTL/algorithm/kernel/Foreach.hpp"
+#include "cuSTL/algorithm/host/Foreach.hpp"
+#include "lambda/Expression.hpp"
+#include <sstream>
+
+#include "cuSTL/algorithm/kernel/Reduce.hpp"
+
+namespace picongpu
+{
+
+template<typename Species>
+ChargeConservation<Species>::ChargeConservation(std::string name, std::string prefix)
+    : name(name), prefix(prefix)
+{
+    Environment<>::get().PluginConnector().registerPlugin(this);
+}
+
+template<typename Species>
+void ChargeConservation<Species>::pluginRegisterHelp(po::options_description& desc)
+{
+    desc.add_options()
+        ((this->prefix + "_frequency").c_str(),
+        po::value<uint32_t > (&this->notifyFrequency)->default_value(0), "notifyFrequency");
+}
+
+template<typename Species>
+std::string ChargeConservation<Species>::pluginGetName() const {return this->name;}
+
+template<typename Species>
+void ChargeConservation<Species>::pluginLoad()
+{
+    Environment<>::get().PluginConnector().setNotificationPeriod(this, this->notifyFrequency);
+}
+
+template<typename ValueType>
+struct Div
+{
+    typedef ValueType result_type;
+
+    template<typename Field>
+    HDINLINE ValueType operator()(Field field) const
+    {
+        const ValueType reciWidth = float_X(1.0) / CELL_WIDTH;
+        const ValueType reciHeight = float_X(1.0) / CELL_HEIGHT;
+        const ValueType reciDepth = float_X(1.0) / CELL_DEPTH;
+        return ((*field(1,0,0)).x() - (*field).x()) * reciWidth +
+               ((*field(0,1,0)).y() - (*field).y()) * reciHeight +
+               ((*field(0,0,1)).z() - (*field).z()) * reciDepth;
+    }
+};
+
+template<typename Species>
+void ChargeConservation<Species>::notify(uint32_t currentStep)
+{
+    /* WARNING: Do not uncomment this. It will cause an internal compiler segmentation fault! */
+    /*namespace vec = PMacc::math;
+    using namespace vec;*/
```

Comment:
Do you know why this will cause an internal compiler segmentation fault? :fearful: